### PR TITLE
Refactor DeployAwsCloudFormationCommand & final cleanup

### PR DESCRIPTION
**Pull request by andyinaus**
As title, just to move away from IConvention installation approach (with the minimum effort).

# How to review this PR

- Ensure convention is removed correctly
- Ensure Factory pattern has properly addressed the `AwsEnvironmentGeneration` issue raised early in our Autofac container registration

**Note:** this is not going to master yet and of course not tested yet.

### CODE REVIEW

**Comment by flin-8** [comment]
For future reference, splitting up your changes into separate commits will makes PR's much easier to review.

For example:
- Renames
- Resharper cleanup
- Autofac/factory change

File: source/Calamari.Aws/UploadAwsS3Command.cs [hunk 1]
```diff
@@ -29,68 +28,76 @@ namespace Calamari.Aws
...
-                throw new CommandException("Could not find package file: " + deployment.PackageFilePath);
+            if (amazonS3Cleint.IsValueCreated) amazonS3Cleint.Value.Dispose();
 
+            base.Dispose();
+        }
+
+        //TODO: Further refactor is necessary if we have the capacity
+        protected override async Task ExecuteCore()
+        {
+            // TODO: these variable names need to be shared with Sashimi.Aws, but how?
```
Comment (flin-8) [inline source/Calamari.Aws/UploadAwsS3Command.cs:71]: https://octopusdeploy.slack.com/archives/C24PH407Q/p1589930762203700

Can be done in a separate PR

File: source/Calamari.Aws/ApplyCloudFormationChangesetCommand.cs [hunk 2]
```diff
@@ -1,38 +1,36 @@
...
+            IAmazonClientFactory amazonClientFactory,
             ICloudFormationService cloudFormationService)
-            : base (log, variables, amazonSecurityTokenService, amazonIdentityManagementService)
+            : base(log, variables, amazonClientFactory)
         {
             this.cloudFormationService = cloudFormationService;
         }
 
-        protected override void Execute(RunningDeployment deployment)
+        protected override async Task ExecuteCore()
```
Comment (benPearce1) [inline source/Calamari.Aws/ApplyCloudFormationChangesetCommand.cs:25]: `ExecuteAsync` might be a better name here

File: source/Calamari.Aws/DeleteCloudFormationCommand.cs [hunk 3]
```diff
@@ -1,38 +1,33 @@
...
             this.cloudFormationService = cloudFormationService;
         }
 
-        protected override void Execute(RunningDeployment deployment)
+        protected override async Task ExecuteCore()
         {
-            var stackArn = new StackArn(deployment.Variables.Get(AwsSpecialVariables.CloudFormation.StackName));
-            var waitForCompletion = ((VariableDictionary)deployment.Variables).EvaluateTruthy(variables.Get("waitForCompletion"));
+            var stackArn = new StackArn(variables.Get(SpecialVariableNames.Aws.CloudFormation.StackName));
+            var waitForCompletion = ((VariableDictionary) variables).EvaluateTruthy(variables.Get(SpecialVariableNames.Action.WaitForCompletion));
```
Comment (benPearce1) [inline source/Calamari.Aws/DeleteCloudFormationCommand.cs:28]: I forgot there is a `variables.GetFlag` method we can use here, you can set the default value to `True`

File: source/Calamari.Aws/DeployAwsCloudFormationCommand.cs [hunk 4]
```diff
@@ -1,143 +1,157 @@
...
+                       ?.Contains(SpecialVariableNames.Aws.CloudFormation.ChangeSets.Feature) ?? false;
         }
 
-        private bool ChangesetsDisabled(RunningDeployment deployment)
+        string GenerateChangeSetName()
         {
-            return !ChangesetsEnabled(deployment);
+            var name = $"octo-{Guid.NewGuid():N}";
+
+            if (((VariableDictionary)variables).EvaluateTruthy(variables.Get(SpecialVariableNames.Aws.CloudFormation.ChangeSets.Generate)))
```
Comment (benPearce1) [inline source/Calamari.Aws/DeployAwsCloudFormationCommand.cs:118]: `GetFlag`

File: source/Calamari.Aws/Deployment/CloudFormation/StackStatus.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Calamari.Aws.Deployment.CloudFormation
```
Comment (benPearce1) [inline source/Calamari.Aws/Deployment/CloudFormation/StackStatus.cs:1]: How were these values represented before?

File: source/Sashimi.sln [hunk 6]
```diff
@@ -41,8 +41,6 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Calamari.Aws.Tests", "Calam
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Sashimi.Aws.Tests", "Sashimi.Aws.Tests\Sashimi.Aws.Tests.csproj", "{43FED2F0-11D1-4AE2-8172-6266B61B4121}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Calamari.Shared", "Calamari.Shared\Calamari.Shared.csproj", "{61AC9D0E-D3B7-47B3-9331-4154F390215D}"
```
Comment (benPearce1) [inline source/Sashimi.sln:44]: 👍

File: source/Calamari.Aws/Deployment/CloudFormation/StackStatus.cs (hunk 5 above)
Comment (andyinaus) [reply source/Calamari.Aws/Deployment/CloudFormation/StackStatus.cs:1]: Was nested in the `DeployAwsCloudFormationConvention.cs`

File: source/Calamari.Aws/ApplyCloudFormationChangesetCommand.cs [hunk 7]
```diff
@@ -1,38 +1,36 @@
...
 using Calamari.Commands.Support;
-using Calamari.Deployment;
 using Octostache;
 
 namespace Calamari.Aws
 {
-    [Command(KnownAwsCalamariCommands.Commands.ApplyAwsCloudformationChangeset, Description = "Apply an existing AWS CloudFormation changeset")]
-    public class ApplyCloudFormationChangeSetCommand: AwsCommand
+    [Command(KnownAwsCalamariCommands.Commands.ApplyAwsCloudFormationChangeSet, Description = "Apply an existing AWS CloudFormation changeset")]
+    public class ApplyCloudFormationChangeSetCommand : AwsCommand
```
Comment (johnsimons) [inline source/Calamari.Aws/ApplyCloudFormationChangesetCommand.cs:11]: Rename file to match class name

File: source/Calamari.Aws/AwsCommand.cs [hunk 8]
```diff
@@ -1,57 +1,74 @@
...
+using Calamari.Aws.Integration.CloudFormation;
+using Calamari.Aws.Util;
 using Calamari.Commands.Support;
-using Calamari.Deployment;
 using Octopus.CoreUtilities.Extensions;
 
 namespace Calamari.Aws
 {
-    public abstract class AwsCommand : ICommand
+    public abstract class AwsCommand : ICommand, IDisposable
```
Comment (johnsimons) [inline source/Calamari.Aws/AwsCommand.cs:18]: Who disposes of `ICommand`?

Comment (johnsimons) [reply source/Calamari.Aws/AwsCommand.cs:18]: Are `ICommand`s registered in the container, I think they are

Comment (johnsimons) [reply source/Calamari.Aws/AwsCommand.cs:18]: All good

File: source/Calamari.Aws/UploadAwsS3Command.cs [hunk 9]
```diff
@@ -267,30 +276,33 @@ Task<S3UploadResult> UploadSingleFileSelection(RunningDeployment deployment, S3S
...
         /// <param name="deployment"></param>
         /// <param name="options"></param>
-        Task<S3UploadResult> UploadUsingPackage(RunningDeployment deployment, S3PackageOptions options)
+        async Task<S3UploadResult> UploadUsingPackage(RunningDeployment deployment, S3PackageOptions options)
         {
             Guard.NotNull(deployment, "Deployment may not be null");
             Guard.NotNull(options, "Package options may not be null");
 
             var filename = GetNormalizedPackageFilename(deployment);
+            var client = await amazonS3Cleint.Value;
```
Comment (johnsimons) [inline source/Calamari.Aws/UploadAwsS3Command.cs:297]: spelling

**Pull request merged**